TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,2695 @@
+        /// <param name="symRef">The Symbol Reference to the Section os Paragraph</param>
+        /// <returns>The Symbol of the section or Paragraph if resolved, null otherwise.</returns>
+        private Symbol CheckSectionOrParagraph(Node node, SymbolReference symRef)
+        {
+            //Resolve the target Section or Paragraph.
+            Container<Symbol>.Entry symbols = ResolveSectionOrParagraphSymbol(symRef);
+
+            if (symbols.Count == 0)
+            {
+                Diagnostic d = new Diagnostic(MessageCode.SemanticTCErrorInParser,
+                    node.CodeElement.Column,
+                    node.CodeElement.Column,
+                    node.CodeElement.Line,
+                    string.Format(Resource.UnknownSectionOrParagraph, symRef.ToString()));
+                Diagnostics.Add(d);
+                return null;
+            }
+            if (symbols.Count > 1)
+            {
+                Diagnostic d = new Diagnostic(MessageCode.SemanticTCErrorInParser,
+                    node.CodeElement.Column,
+                    node.CodeElement.Column,
+                    node.CodeElement.Line,
+                    string.Format(Resource.AmbiguousSectionOrParagraph, symRef.ToString()));
+                Diagnostics.Add(d);
+                return null;
+            }
+
+            return symbols.Symbol;
+        }
+
+        /// <summary>
+        /// Store all procedure's sentence blocks in a group.
+        /// </summary>
+        /// <param name="p">The procedure node</param>
+        /// <param name="procedureSymbol">The procedure symbol</param>
+        /// <param name="group">The Group in which to store all blocks.</param>
+        /// <param name="clonedBlockIndexMap">The Map of cloned map indices from the original indices to the new indices of block</param>
+        private void StoreProcedureSentenceBlocks(PerformProcedure p, Symbol procedureSymbol, BasicBlockForNodeGroup group, Dictionary<int, int> clonedBlockIndexMap)
+        {
+            IEnumerable<CfgSentence> procedureSentences = YieldSectionOrParagraphSentences(procedureSymbol);
+            foreach (var sentence in procedureSentences)
+            {
+                //A Sentence has at least one block
+                System.Diagnostics.Debug.Assert(sentence.AllBlocks != null);
+                System.Diagnostics.Debug.Assert(sentence.AllBlocks.First.Value == sentence.Block);
+                foreach (var block in sentence.AllBlocks)
+                {//We must clone each block of the sequence and add them to the group.                    
+                    //System.Diagnostics.Debug.Assert(!clonedBlockIndexMap.ContainsKey(block.Index));
+                    if (!clonedBlockIndexMap.ContainsKey(block.Index))
+                    {//If this block has been already add, this mean there are recursive GOTOs
+                        clonedBlockIndexMap[block.Index] = block.Index;
```

Comment:
This is the only add operation on the dictionary so the keys are always equal to the values. The Dictionary can be turned into a List.

Reply:
No, looking in the list as a complexity of O(n) where n is the count of element in the list, but looking in a Dictionary as a complexity of O(1) method _ContainsKeys But List Contains method as a complexcity of O(n)

Reply:
Ok then we should use `HashSet<int>` which has a `Contains` method with a complexity of O(1).

Reply:
ok

Reply:
done